lib/rjgit.rb

```diff
@@ -209,15 +210,38 @@ def self.grep(repository, query, options={})
 
       query = Regexp.new(query.source, query.options | Regexp::IGNORECASE) if case_insensitive
 
-      ls_tree(repo, nil, options.fetch(:ref, 'HEAD'), ls_tree_options).each_with_object({}) do |item, result|
-        blob = Blob.new(repo, item[:mode], item[:path], walk.lookup_blob(ObjectId.from_string(item[:id])))
-        next if blob.binary?
+      # We optimize below by first grepping the entire file, and then, if a match is found, then identifying the individual line.
+      # To avoid missing full-line matches during the optimization, we first convert multiline anchors to single-line anchors.
+      query = Regexp.new(query.source.gsub(/\A\\A/, '^').gsub(/\\z\z/, '$'), query.options)
 
-        rows = blob.data.split("\n")
+      ref = options.fetch(:ref, 'HEAD')
+      files_to_scan = ls_tree(repo, nil, ref, ls_tree_options)
+
+      files_to_scan.each_with_object({}) do |file, result|
+        id = if file[:mode] == SYMLINK_TYPE
+          symlink_source = repo.open(ObjectId.from_string(file[:id])).get_bytes.to_a.pack('c*').force_encoding('UTF-8')
+          unless symlink_source[File::SEPARATOR]
+            dir = file[:path].split(File::SEPARATOR)
+            dir[-1] = symlink_source
+            symlink_source = File.join(dir)
+          end
+          Blob.find_blob(repo, symlink_source, ref).jblob.id
+        else
+          ObjectId.from_string(file[:id])
+        end
+        bytes = repo.open(id).get_bytes
+
+        binary = RawText.is_binary(bytes)
+        next if binary
```

Comment:
Perhaps just `next if RawText.is_binary(bytes)`, to avoid instantiating another object. We're not using `binary` anywhere else, right?

Reply:
Seems like it.

Reply:
Fixed: d7b6c75e1c6fa7ad7dc6a30c0e1740ab8b773b2a